### DISCUSSION TITLE
Renamer transformer improvement suggestions

### DISCUSSION THREAD

**Discussion by damienwojtowicz** [online-ml/river#820] [Ideas]
Hi all,

Currently, there is no way to rename dictionary keys within a pipeline without resorting to a custom function included in a `FuncTransformer`, since `Renamer` transformer only allows for prefixing or suffixing dictionary keys. I think `Renamer` should behave more like [panda's rename function for DataFrames](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html), so users can define key renaming rules as a `Dict[str, str]`. It would behave as follows :

```python
>>> x = {'foo': 42, 'bar': 24}
>>> renamer = Renamer({'foo' : 'baz', 'bar': 'dummy'})
>>> renamer.transform_one(x)
{'baz': 42, 'dummy': 24}
```
Obviously, prefixing and suffixing should be preserved (and perhaps extended by letting users define collections of keys that should not be affected? or maybe picking the one they wish to alter?). I suggest to separate those in a `Prefixer` and `Suffixer` transformers :
```python
>>> x = {'foo':42, 'bar':24}
>>> pipeline = Renamer({'foo': 'baz'}) 
...     | Prefixer('x_', exclude={'bar'}) 
...     | Suffixer('_in', only={'baz'})
>>> pipeline.transform_one(x)
{'x_baz_in': 42, 'bar': 24}
```
What do you think about it?

**Comment by MaxHalford** [comment]
Thanks @damienwojtowicz, I like this proposal!

I think @AdilZouitine had some thoughts on this. Adil, can I let you own this one?

**Comment by AdilZouitine** [comment]
Hi @damienwojtowicz, 
I had the same idea as you a few weeks ago. 
The Renamer API that you propose seems to be very good. 

@MaxHalford  yes!

**Comment by MaxHalford** [comment]
I would suggest thinking about what happens when a feature is missing. You might also want to add a unit test for that :)

**Comment by damienwojtowicz** [comment]
@AdilZouitine I'll take a look and make a PR soon then (I'm thinking about starting to contribute to river), if you're OK with it. Following what Max said, I believe missing features should be ignored (or maybe optionally raise a warning for debugging). It's more flexible IMO.

**Reply by AdilZouitine** [reply]
Great @damienwojtowicz , I will be happy to review your pull request. 
I think you should ignore the missing features. 
However, it will be important to write this choice in the documentation.

**Reply by MaxHalford** [reply]
What about the `exclude` and `only` parameters? Usually I would handle that functionality `compose.Select` and `compose.Discard`. I suggest because I don't like to have two ways of doing things. But I'll let you make the final decision :)

**Reply by damienwojtowicz** [reply]
You're right, the parameters are redundant with `Select` and `Discard`  :see_no_evil: Forget about them!
